jans-cedarling/cedarling/src/authz/entities/mod.rs

```diff
@@ -33,22 +33,49 @@ pub(crate) struct AccessTokenEntities {
 
 impl AccessTokenEntities {
     /// Map all values to vector of cedar-policy entities
-    pub fn entities(self) -> Vec<cedar_policy::Entity> {
-        vec![self.workload_entity, self.access_token_entity]
+    pub fn into_iter(self) -> impl Iterator<Item = cedar_policy::Entity> {
+        vec![self.workload_entity, self.access_token_entity].into_iter()
     }
 }
 
 /// Create all entities from AccessToken
 pub fn create_access_token_entities(
     schema: &CedarSchemaJson,
-    data: &TokenPayload,
+    data: &AccessTokenData,
 ) -> Result<AccessTokenEntities, AccessTokenEntitiesError> {
     Ok(AccessTokenEntities {
-        access_token_entity: meta::AccessTokenMeta.create_entity(schema, data)?,
-        workload_entity: meta::WorkloadEntityMeta.create_entity(schema, data)?,
+        access_token_entity: meta::AccessTokenMeta.create_entity(schema, &data.0)?,
+        workload_entity: meta::WorkloadEntityMeta.create_entity(schema, &data.0)?,
     })
 }
 
+/// Create id_token entity
+pub fn id_token_entity(
+    schema: &CedarSchemaJson,
+    data: &IdTokenData,
+) -> Result<cedar_policy::Entity, CedarPolicyCreateTypeError> {
+    meta::IdToken.create_entity(schema, &data.0)
+}
+
+/// Create user entity
+pub fn user_entity(
+    schema: &CedarSchemaJson,
+    id_token_data: &IdTokenData,
+    userinfo_token_data: &UserInfoTokenData,
+) -> Result<cedar_policy::Entity, CedarPolicyCreateTypeError> {
+    const SUB_KEY: &str = "sub";
+
+    // if 'sub' is not the same we discard the userinfo token
+    let data = if id_token_data.0.payload.get(SUB_KEY) == userinfo_token_data.0.payload.get(SUB_KEY)
+    {
+        &id_token_data.0.extend(userinfo_token_data.0.clone())
+    } else {
+        &id_token_data.0
+    };
+
+    meta::User.create_entity(schema, data)
+}
```

Comment:
To avoid the cloning in `token_data.rs`
```suggestion
pub fn user_entity(
    schema: &CedarSchemaJson,
    id_token_data: &mut IdTokenData,
    userinfo_token_data: &UserInfoTokenData,
) -> Result<cedar_policy::Entity, CedarPolicyCreateTypeError> {
    const SUB_KEY: &str = "sub";

    // If 'sub' matches in both tokens, extend id_token_data with userinfo_token_data
    if id_token_data.0.payload.get(SUB_KEY) == userinfo_token_data.0.payload.get(SUB_KEY) {
        id_token_data.0.extend(userinfo_token_data.0.clone()); // Mutate in place
    }

    meta::User.create_entity(schema, &id_token_data.0)
}
```

Reply:
`id_token_data: &mut IdTokenData,`

Using mutable reference here can lead to bugs in the future.
It is bad practice and premature optimization.

`Value object` should stay immutable.

Reply:
I see your point of view.
Maybe you can just change `.0` by a getter to have a code more readable.  
```
impl IdTokenData {
    pub fn token_payload(&self) -> &TokenPayload {
        &self.0
    }
}
```

Reply:
implemented with derive implementation